core/kroma_migration.go

```diff
@@ -0,0 +1,197 @@
+	destructChangesPrefix = []byte("dC-")
+	migratedRootKey       = []byte("MigratedRoot")
+	migratedNumberKey     = []byte("MigratedNumber")
+)
+
+func NewMigratedRef(db ethdb.Database) *MigratedRef {
+	ref := &MigratedRef{db: db}
+	if b, _ := db.Get(migratedRootKey); len(b) > 0 {
+		ref.root = common.BytesToHash(b)
+	}
+	if b, _ := db.Get(migratedNumberKey); len(b) > 0 {
+		ref.number = hexutil.MustDecodeUint64(string(b))
+	}
+	return ref
+}
+
+func (mr *MigratedRef) Update(root common.Hash, number uint64) error {
+	mr.mu.Lock()
+	defer mr.mu.Unlock()
+
+	mr.root = root
+	mr.number = number
+
+	if err := mr.db.Put(migratedRootKey, root.Bytes()); err != nil {
+		return fmt.Errorf("failed to update migrated state root: %w", err)
+	}
+	if err := mr.db.Put(migratedNumberKey, []byte(hexutil.EncodeUint64(number))); err != nil {
+		return fmt.Errorf("failed to update migrated block number: %w", err)
+	}
+	return nil
+}
+
+func (mr *MigratedRef) Root() common.Hash {
+	mr.mu.Lock()
+	defer mr.mu.Unlock()
+	return mr.root
+}
+
+func (mr *MigratedRef) BlockNumber() uint64 {
+	mr.mu.Lock()
+	defer mr.mu.Unlock()
+	return mr.number
+}
+
+// encodeBlockNumber encodes a block number as big endian uint64
+func encodeBlockNumber(number uint64) []byte {
+	enc := make([]byte, 8)
+	binary.BigEndian.PutUint64(enc, number)
+	return enc
+}
+
+func AccountChangesKey(blockNumber uint64) []byte {
+	return append(accountChangesPrefix, encodeBlockNumber(blockNumber)...)
+}
+
+func StorageChangesKey(blockNumber uint64) []byte {
+	return append(storageChangesPrefix, encodeBlockNumber(blockNumber)...)
+}
+
+func DestructChangesKey(blockNumber uint64) []byte {
+	return append(destructChangesPrefix, encodeBlockNumber(blockNumber)...)
+}
+
+func ReadStateChanges(db ethdb.KeyValueStore, blockNumber uint64) (map[common.Address]bool, map[common.Hash][]byte, map[common.Hash]map[common.Hash][]byte, error) {
+	enc, err := db.Get(DestructChangesKey(blockNumber))
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	destruct, err := DeserializeStateChanges[map[common.Address]bool](enc)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	enc, err = db.Get(AccountChangesKey(blockNumber))
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	accounts, err := DeserializeStateChanges[map[common.Hash][]byte](enc)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	enc, err = db.Get(StorageChangesKey(blockNumber))
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	storages, err := DeserializeStateChanges[map[common.Hash]map[common.Hash][]byte](enc)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	return destruct, accounts, storages, nil
+}
+
+// Note: it's should be called in MigrationTime
+func WriteStateChanges(db ethdb.KeyValueStore, blockNumber uint64, stateObjectsDestruct map[common.Address]*types.StateAccount, accounts map[common.Hash][]byte, storages map[common.Hash]map[common.Hash][]byte) error {
+	batch := db.NewBatch()
+	stateObjectsDestructInfo := make(map[common.Address]bool)
```

Comment:
Good question. 
- The account key and storage key are zk-hashed addresses. After statedb.Commit() is completed and the preimages are stored, these preimages can then be retrieved by the migrator.
- In some cases, the stateObjects in stateObjectDestruct don't exist in the DB (in other words, their preimages are not stored). Therefore, we store the preimage instead of the hashed address.

Reply:
Oh I mean can't we do like this?
```go
	var stateObjectsDestructInfo []common.Address

	for addr := range stateObjectsDestruct {
		if addr.Cmp(params.SystemAddress) == 0 {
			continue
		}
		stateObjectsDestructInfo = append(stateObjectsDestructInfo, addr)
	}
```

Reply:
I got it. It's nothing special, I just followed the data structure used by Geth. Maybe using a slice would be more efficient.

Reply:
@Pangssu What do you think about using a slice instead of a map?

Reply:
I think it would be better to just save as `map[common.Address]*types.StateAccount`. 
Wouldn't it be better to save the original data as it is?

Reply:
stateObjectsDestructInfo is only used for deleting the account, so it's just enough to store boolean that indicates to delete it or not, or just store the list of accounts in a slice as I suggested. @Pangssu Why do you think it's better to save the original data?

Reply:
I thought it would be better to just store the **pure data** without any changes. However, it will use more storage.